### PR TITLE
Add RationalQuadraticSpline

### PR DESCRIPTION
**Pull request by gvcallen**
Adds the rational quadratic spline bijector, very useful for complex normalizing flows. Reference distrax code is [here](https://github.com/google-deepmind/distrax/blob/main/distrax/_src/bijectors/rational_quadratic_spline.py)

### CODE REVIEW

File: distreqx/bijectors/_rational_quadratic_spline.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    x_pos_bin = (params_bin_left[0], params_bin_right[0])
+    y_pos_bin = (params_bin_left[1], params_bin_right[1])
+    knot_slopes_bin = (params_bin_left[2], params_bin_right[2])
+
+    bin_width = x_pos_bin[1] - x_pos_bin[0]
+    bin_height = y_pos_bin[1] - y_pos_bin[0]
+    bin_slope = bin_height / bin_width
+
+    z = (x - x_pos_bin[0]) / bin_width
+    z = jnp.clip(z, 0.0, 1.0)
```
Comment (lockwo) [inline distreqx/bijectors/_rational_quadratic_spline.py:96]: this function has some good comments in the distrax codebase, we should copy them over

File: distreqx/bijectors/_rational_quadratic_spline.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+class RationalQuadraticSpline(
+    AbstractFowardInverseBijector,
+    AbstractInvLogDetJacBijector,
+    AbstractFwdLogDetJacBijector,
+    strict=True,
+):
+    """A rational-quadratic spline bijector.
+
+    Implements the spline bijector introduced by:
+    > Durkan et al., Neural Spline Flows, https://arxiv.org/abs/1906.04032, 2019.
```
Comment (lockwo) [inline distreqx/bijectors/_rational_quadratic_spline.py]: we can do the special bib notation here for mkdocs (e.g. ??? cite "References")

File: distreqx/bijectors/_rational_quadratic_spline.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+        params: Array,
+        range_min: float,
+        range_max: float,
+        boundary_slopes: str = "unconstrained",
+        min_bin_size: float = 1e-4,
+        min_knot_slope: float = 1e-4,
+    ):
+        """Initializes a RationalQuadraticSpline bijector.
+
+        Args:
```
Comment (lockwo) [inline distreqx/bijectors/_rational_quadratic_spline.py]: should match docstring style (e.g. `**Arguments**`)

File: distreqx/bijectors/_rational_quadratic_spline.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            )
+
+        self.num_bins = (params.shape[-1] - 1) // 3
+        dtype = params.dtype
+
+        unnormalized_bin_widths = params[..., : self.num_bins]  # noqa: E203
+        unnormalized_bin_heights = params[
+            ..., self.num_bins : 2 * self.num_bins  # noqa: E203
+        ]
+        unnormalized_knot_slopes = params[..., 2 * self.num_bins :]  # noqa: E203
```
Comment (lockwo) [inline distreqx/bijectors/_rational_quadratic_spline.py:344]: I thought I fixed the black/flake interaction for these, does it error if you remove the `# noqa: E203 `?

File: tests/rational_quadratic_spline_test.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+from distreqx.bijectors import RationalQuadraticSpline
+
+
+class RationalQuadraticSplineTest(TestCase):
```
Comment (lockwo) [inline tests/rational_quadratic_spline_test.py:11]: There's a few more low hanging fruit from the distrax tests, e.g. https://github.com/google-deepmind/distrax/blob/main/distrax/_src/bijectors/rational_quadratic_spline_test.py#L153 that are worth porting over

**Review by lockwo — APPROVED** [review]
